### PR TITLE
Move kernel defines in zfs_znode.h to platform code.

### PR DESCRIPTION
**Pull request by mattmacy**
The ones used by FreeBSD are quite different.

Signed-off-by: Matt Macy <mmacy@FreeBSD.org>

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by lundman** [comment]
We are moving the definition of `struct znode` to platform code?

**Comment by mattmacy** [comment]
> We are moving the definition of `struct znode` to platform code?

Yes.
This is the difference between the  Linux and the FreeBSD versions
http://dpaste.com/37BRKC3

**Comment by behlendorf** [comment]
We can always further refactor this and pull out more common bits if it's beneficial.  But this looks like a good place to start.

**Comment by lundman** [comment]
So we are finally here, eh.  Tongue-in-cheek people, here we go...

The diff didn't look particularly scary, the type change between uint and boolean seems mostly a preference thing, I don't mind either way. But all the ZFS code will not work if the members of `znode` aren't present, so it "feels" like a mistake to let each platform have "a go" at their own definition.

But sure, there is that "local filesystem object pointer", either `inode` or `vnode` that seems pretty straight forward. Even though they are the same concept, just a different name. One hand it feels pedantic to insist it be called `inode` when `vnode` is a perfectly cromulent name, but why should one platform "be less" - so then, why call it `inode` or `vnode` at all - could just be `fsnode` or something made up, then all platforms are "wrong". You should see the name on Windows, hah! 

 But we finally get to decide what to do about `ZTOV` macros, is it worth renaming everything to `ZTOI` ? Should it be a `ZTOL` for "local type". This we will see around the code after all. Really, I should have added something like `ZTOZSB` from the beginning, instead of `zp->z_zfsvfs->z_vfs` - but missed opportunity, that macros is running unopposed so it's a de-facto winner!

:)

**Comment by behlendorf** [comment]
This PR could definitely use some additional context!  It's by no means the final version of anything, just one more proposed step we want to get feedback on.  Let me fill everyone in on some additional details about what the current thinking is.  Then please feel free to poke holes in why it won't work, or is a terrible idea.  :)

The high level design proposal we've been moving towards has a few key elements, but it broadly goes like this (this PR includes some, but not all, of this).

* Only `znode_t`'s will be used / passed around in the common code.
* A single shared `znode_t` structure defined in `sys/zfs_znode.h` is highly desirable, plus as much additional common code as possible.  That looks doable with a little more refactoring.
* Each platform provides a `zfs_znode_impl.h` with the following as appropriate:
  * `ZTOI / ZTOV`, `ZTOZSB / ZTOVFS`, etc: These can be used in platform specific code.  Every platform then gets something which makes sense.
  * Add `zget(znode_t *)` and `zput(znode_t *)`:  Wrappers for `ihold() / vhold()` and `iput()  / vrele()` which can be used in the common code.
  * Possibly the `ZFS_ENTER / ZFS_EXIT` and related macros.
  * Any platform specific function prototypes from `znode.c`.
  * Whatever else was missed.

Does this sound reasonable?  Would it be helpful to iterate in this PR until we address most/all of the point above.  Or shall we split this over a couple PRs.

**Comment by ikozhukhov** [comment]
we have no ZTOI (inode version) on illumos/dilos too and using ZTOV.
will be much more better prepare wrap fro it where it possible.

**Comment by mattmacy** [comment]
I’m really trying to get the number of common files changed down to zero so I can push FreeBSD support by the end of this year. I’ve already gone through and refactored things like zvol to share as much as possible, but refactoring znode feels like a refinement. Many headers and .c files will ultimately have a finer grained split.

**Comment by mattmacy** [comment]
> we have no ZTOI (inode version) on illumos/dilos too and using ZTOV.
> will be much more better prepare wrap fro it where it possible.

**Comment by mattmacy** [comment]
> So we are finally here, eh. Tongue-in-cheek people, here we go...
> 
> The diff didn't look particularly scary, the type change between uint and boolean seems mostly a preference thing, I don't mind either way. But all the ZFS code will not work if the members of `znode` aren't present, so it "feels" like a mistake to let each platform have "a go" at their own definition.
> 
> But sure, there is that "local filesystem object pointer", either `inode` or `vnode` that seems pretty straight forward. Even though they are the same concept, just a different name. One hand it feels pedantic to insist it be called `inode` when `vnode` is a perfectly cromulent name, but why should one platform "be less" - so then, why call it `inode` or `vnode` at all - could just be `fsnode` or something made up, then all platforms are "wrong". You should see the name on Windows, hah!
> 
> But we finally get to decide what to do about `ZTOV` macros, is it worth renaming everything to `ZTOI` ? Should it be a `ZTOL` for "local type". This we will see around the code after all. Really, I should have added something like `ZTOZSB` from the beginning, instead of `zp->z_zfsvfs->z_vfs` - but missed opportunity, that macros is running unopposed so it's a de-facto winner!
> 
> :)

@lundman I've made much more common. I'm sure there's still room for improvement, but this is a first pass.

File: include/os/linux/zfs/sys/zfs_znode_impl.h [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+ *    freed etc and zp_size reset. zp_size must be set within range lock.
+ * 2. For writes and punching holes (zfs_write & zfs_space) just the range
+ *    being written or freed needs to be locked as RL_WRITER.
+ *    Multiple writes at the end of the file must coordinate zp_size updates
+ *    to ensure data isn't lost. A compare and swap loop is currently used
+ *    to ensure the file size is at least the offset last written.
+ * 3. For reads (zfs_read, zfs_get_data & zfs_putapage) just the range being
+ *    read needs to be locked as RL_READER. A check against zp_size can then
+ *    be made for reading beyond end of file.
+ */
```
Comment (behlendorf) [inline include/os/linux/zfs/sys/zfs_znode_impl.h]: To the best of my knowledge these rules apply to all of the platforms.  Plus this no longer seems like the best place to document them.  I'd suggest we go ahead and move them to the top of `zfs_rlock.c`.  While the range lock implementation is generic, it also happens to only be used for this purpose.  Including the rules here, along with the rest of the implementation seems like the most logical shared location.

```c
 * Interface
 * ---------
 * Defined in zfs_rlock.h but essentially:
 *      lr = rangelock_enter(zp, off, len, lock_type);
 *      rangelock_reduce(lr, off, len); // optional
 *      rangelock_exit(lr);
 *
 * Range locking rules
 * --------------------
 * 1. When truncating a file (zfs_create, zfs_setattr, zfs_space) the whole
 *    file range needs to be locked as RL_WRITER. Only then can the pages be
 *    freed etc and zp_size reset. zp_size must be set within range lock.
 *
 * 2. For writes and punching holes (zfs_write & zfs_space) just the range
 *    being written or freed needs to be locked as RL_WRITER.  Multiple
 *    writes at the end of the file must coordinate zp_size updates to ensure
 *    data isn't lost. A compare and swap loop is currently used to ensure
 *    the file size is at least the offset last written.
 *
 * 3. For reads (zfs_read, zfs_get_data & zfs_putapage) just the range being
 *    read needs to be locked as RL_READER. A check against zp_size can then
 *    be made for reading beyond end of file.
 *
```

File: include/os/linux/zfs/sys/zfs_znode_impl.h [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+#define	ZTOTYPE(zp)	(ZTOI(zp)->i_mode)
+#define	ZTOGID(zp) (ZTOI(zp)->i_gid)
+#define	ZTOUID(zp) (ZTOI(zp)->i_uid)
+#define	ZTONLNK(zp) (ZTOI(zp)->i_nlink)
+
+#define	Z_ISBLK(type) S_ISBLK(type)
+#define	Z_ISCHR(type) S_ISCHR(type)
+#define	Z_ISLNK(type) S_ISLNK(type)
+#define	S_ISDEV(type)	(S_ISCHR(type) || S_ISBLK(type) || S_ISFIFO(type))
```
Comment (behlendorf) [inline include/os/linux/zfs/sys/zfs_znode_impl.h]: S_ISDEV -> Z_ISDEV while we're here?

File: include/os/linux/zfs/sys/zfs_znode_impl.h [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+#ifdef	__cplusplus
+extern "C" {
+#endif
+
+#define	ZNODE_OS_FIELDS			\
+	boolean_t	z_is_mapped;		\
+	boolean_t	z_is_ctldir;		\
+	boolean_t	z_is_stale;		\
+	boolean_t	z_suspended;		\
```
[the remark this answers is not in the document]
Comment (mattmacy) [reply include/os/linux/zfs/sys/zfs_znode_impl.h]: So long as it doesn't conflict with anything on FreeBSD that's fine with. ZFS structures aren't exactly optimized with an eye for cache line access patterns (i.e. the more the merrier).

File: module/zfs/zfs_rlock.c [hunk 4]
```diff
@@ -99,6 +99,28 @@
 #include <sys/zfs_context.h>
 #include <sys/zfs_rlock.h>
 
+
+/*
+ * Defined in zfs_rlock.h but essentially:
+ *      lr = rangelock_enter(zp, off, len, lock_type);
+ *      rangelock_reduce(lr, off, len); // optional
+ *      rangelock_exit(lr);
```
Comment (behlendorf) [inline module/zfs/zfs_rlock.c]: Up a bit higher, this exact block appears starting on line 36

**Pull request merged**